go-selinux/selinux_linux.go

```diff
@@ -1095,3 +1104,115 @@ func dupSecOpt(src string) ([]string, error) {
 func disableSecOpt() []string {
 	return []string{"disable"}
 }
+
+func getDefaultContextFromReaders(c defaultSECtx) (string, error) {
+	if c.verifier == nil {
+		return "", fmt.Errorf("verifier function must not be nil")
+	}
+
+	context, err := newContext(c.scon)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create label for %s", c.scon)
+	}
+
+	findUserInContext := func(r io.Reader) string {
+		fromRole := context["role"]
+		fromType := context["type"]
+		bufin := bufio.NewReader(r)
+
+		for done := false; !done; {
+			var (
+				line string
+				err  error
+			)
+
+			if line, err = bufin.ReadString('\n'); err != nil {
+				if err == io.EOF {
+					done = true
+				} else {
+					break
+				}
+			}
+			line = strings.TrimSpace(line)
+			if len(line) == 0 {
+				// Skip blank lines
+				continue
+			}
+			if line[0] == ';' || line[0] == '#' {
+				// Skip comments
+				continue
+			}
+
+			conns := strings.Fields(line)
+
+			lineConn := strings.TrimSpace(conns[0])
+			lineArr := strings.SplitN(lineConn, ":", 3)
+
+			// malformed context in the context file
+			if len(lineArr) != 3 {
+				continue
+			}
+
+			if lineArr[0] != fromRole || lineArr[1] != fromType {
+				continue
+			}
+
+			for _, cc := range conns[1:] {
+				userConn := strings.TrimSpace(cc)
+				userArr := strings.SplitN(userConn, ":", 3)
+
+				context["user"] = c.user
+				context["role"] = userArr[0]
+				context["type"] = userArr[1]
+				context["level"] = c.level
+
+				outConn := context.get()
+				if err := c.verifier(outConn); err != nil {
+					continue
+				}
+
+				return outConn
```

Comment:
Overall I think this should return errors, rather than silently ignore all input it can't process.

Reply:
Or some of it... I dunno

Reply:
In the libselinux implementation, they loop until a valid match is found, silently ignoring the invalid contexts. Maybe it is the right thing to do to return an error if the entry is invalid, since the user should be aware of invalid contexts, but we lose feature parity with libselinux. I defer to you on what you want me to do.

Reply:
The idea in libselinux is that if there's a typo in one context, it is better to skip the wrong entry and try to parse the rest.